include/dynamatic/Transforms/Speculation/SpeculationPlacement.h

```diff
@@ -0,0 +1,108 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// This file declares the data structures needed for speculation algorithms
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef DYNAMATIC_TRANSFORMS_SPECULATION_PLACEMENT_H
+#define DYNAMATIC_TRANSFORMS_SPECULATION_PLACEMENT_H
+
+#include "circt/Dialect/Handshake/HandshakeOps.h"
+#include "dynamatic/Support/DynamaticPass.h"
+#include "dynamatic/Support/LLVM.h"
+#include "dynamatic/Support/Logging.h"
+#include "dynamatic/Support/TimingModels.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/Hashing.h"
+#include <map>
+#include <string>
+#include <unordered_set>
+#include <vector>
+
+namespace dynamatic {
+namespace speculation {
+
+struct PlacementNames {
+  StringRef srcOpName;
+  StringRef dstOpName;
+};
+
+struct OpPlacement {
+  Value srcOpResult;
+  Operation *dstOp;
+
+  bool operator==(const OpPlacement &other) const;
+
+  struct Hash {
+    std::size_t operator()(const OpPlacement &p) const;
+  };
+};
+
+typedef std::unordered_set<OpPlacement, OpPlacement::Hash> PlacementList;
+
+class SpeculationPlacements {
+private:
+  OpPlacement speculator;
+  PlacementList saves;
+  PlacementList commits;
+  PlacementList saveCommits;
+
+public:
+  // Empty constructor
+  SpeculationPlacements(){};
+
+  // Initializer with source and destination operations for the Speculator
+  SpeculationPlacements(Value srcOpResult, Operation *dstOp) {
+    speculator = {srcOpResult, dstOp};
+  }
+
+  // Set the speculator operations positions according to a JSON file
+  static LogicalResult readFromJSON(std::string &jsonPath,
+                                    SpeculationPlacements &place,
+                                    NameAnalysis &nameAnalysis);
+
+  // Explicitly set the speculator position
+  void setSpeculator(Value srcOpResult, Operation *dstOp);
+
+  // Add the position of a Save operation
+  void addSave(Value srcOpResult, Operation *dstOp);
+
+  // Add the position of a Save operation
+  void addCommit(Value srcOpResult, Operation *dstOp);
+
+  // Add the position of a Save operation
+  void addSaveCommit(Value srcOpResult, Operation *dstOp);
+
+  // Check if there is a commit from srcOp to dstOp
+  bool containsCommit(Value srcOpResult, Operation *dstOp);
+
+  // Check if there is a save from srcOp to dstOp
+  bool containsSave(Value srcOpResult, Operation *dstOp);
+
+  // Remove a commit from the commit placement map
+  void eraseCommit(Value srcOpResult, Operation *dstOp);
+
+  // Remove a save from the save placement map
+  void eraseSave(Value srcOpResult, Operation *dstOp);
+
+  // Get the Placement instance that specifies the Speculator position
+  OpPlacement getSpeculatorPlacement();
+
+  // Get a vector of the existing Save operation placements
+  template <typename T>
```

Comment:
In this case, a general templated function is never defined, but rather only three specializations, and the three should be public. I think that in this case, it would be best to leave the specializations in the .cpp. Otherwise, how would you define these in the .h file?

Reply:
NB: After looking online I think it would in fact work as-is compile/link-wise, but my last point stands nonetheless.

Reply:
NB 2: Just realized that `placeUnits<T>` requires that `getPlacements<T>` is templated too to work elegantly... I understand now why you did it this way. Still not super happy with getPlacements<T> being templated due to the reason above but feel free to ignore all of this then, sorry for the endless rant.

Reply:
I understand your point, thanks for the detailed reasoning. At first, I had created 3 non-templated getter methods. However, as you noticed, when writing `placeUnits<T>` I thought about the possibility of creating a templated getter. I will leave it like this for now, we can rethink about it in the following PRs.